47

darrenklein opened the pull request:
@franko-franicevich  @alhambra1 @bernardd This PR has my proposed fix for issue #47 - in a nutshell, if an `in` query's parameters are exclusive to the primary key field (or fields, in the case of a composite key), then Dynamo's `BatchGetItem` will be called.  For example:

```
from(p in Person, where: p.id in ^ids)
```

or

```
from(bp in BookPage,
   where: bp.id in ^ids
     and bp.page_num in ^pages)
```

However, if the query's parameters include fields that are not exclusive to the primary key's fields, then a filter condition is present and a Dynamo query will be used instead. For example:

```
from(p in Person,
  where: p.id in ^ids
    and is_nil(p.age))
```

I'm curious about your thoughts on my use of anonymous functions within the call to `Query.get_item/3` for `do_fetch_recursive` and `do_in_query`. Any reasons that doesn't seem like a good idea? It'd be easy enough to move them out into private methods if that seemed like a better approach.

Thanks!

alhambra1 commented:
If we can show that with 1000 records, we get a _significant_ improvement in either or both (1) time of execution and (2) overall network bandwidth use, I would agree with the idea of using queries (note that (1) is most advantageous when implementing the queries in different threads, AKA parallelising). Otherwise, batch-get-item is DynamoDBs performant go-to method for retrieving multiple records with known complete record-IDs and we could just have the server filter after receiving the data. Also consider how many of the 1000 records would be expected to be filtered out, when testing.

alhambra1 commented:
In general, I think it's weird to query on a DynamoDB record for which we have a complete record ID for. I'm on the fence in terms of this use case in general.

alhambra1 commented:
Users may be expecting the adapter to provide a general-use API. However, knowledgeable users of DynamoDB, should be expecting the adapter to behave in the smartest way regarding performance.

darrenklein commented:
Yeah, I was wondering about the possibility of using `get-item` and `batch-get-item` and then post-filtering instead - as you note, this would clearly need some testing, but I think that's an interesting approach.

That said, I disagree that it's weird to query when we have the full id - it would be weird if a query was used where only a full id was presented, but if there are filtering conditions, I think this would be the expected behavior - to your last point, I would think that a knowledgeable user of DDB who queried on a primary key with a filter condition would understand that that would be handled via a query and not via `get-item`/`batch-get-item` (whether or not the adapter should do that is beside the point - I just mean that I don't think it's weird or unexpected).

darrenklein commented:
(Don't take my disagreement on that point to mean that I'm not in favor of the solution you proposed - I'm all for investigating that further)

bernardd commented on lib/ecto_adapters_dynamodb/query.ex:
```diff
@@ -46,8 +48,19 @@ defmodule Ecto.Adapters.DynamoDB.Query do
+
+      Enum.reduce(hash_values, response_map, fn(hash_value, acc) ->
+        # When receiving a list of values to query on, construct a custom query for each of those values to pass into do_fetch_recursive/1.
+        %{"Items" => items} = Kernel.put_in(query, [:expression_attribute_values, :hash_key], hash_value)
```
`Kernel` is a magic namespace which is always imported - you don't need to specify it on calls to its functions. In other words, you can just call `put_in` rather than `Kernel.put_in`.

darrenklein replied:
Yeah, I'm familiar with that - I just opted to be explicit. It seems like this overall logic may get revisited - I'd consider just going with the more customary `put_in`.

alhambra1 commented:
Does "you can" mean you prefer it won't be? (I wouldn't but I don't mind if
others do.)

alhambra1 commented:
*"don't need"

>

bernardd commented:
In general I wouldn't include `Kernel` on the calls. There's a couple of special cases where it's necessary (for example if you want to use the pipe operator to feed into the `==` function you have to use `|> Kernel.==(operand)`, but otherwise the general de facto standard is to leave it off.

alhambra1 commented:
In general, I do not feel compelled to adhere to a standard, although probably write some code that adheres to some standard.

bernardd commented:
> I'm curious about your thoughts on my use of anonymous functions within the call to Query.get_item/3 for do_fetch_recursive and do_in_query. Any reasons that doesn't seem like a good idea?

The main reason not to do it in that instance is that `get_item` is already a pretty big function. As a rule of thumb I try to keep functions able to fit on one screen as an absolute maximum (using the entirely unscientific measure of what fits on my personal screen which is about 60 lines). Anything more and you lose the information of what the function does in the detail of how it does it. Don't be afraid to break functions up into bite-sized pieces. If you're indenting more than a couple of levels, you can probably break those indented blocks out into their own funcitons.

darrenklein commented:
That's a good suggestion - although `get_item` was pretty much already about this large prior to this PR, those two anonymous functions were already more or less part of the logic of `get_item`. Regardless - good opportunity to make it more readable.

bernardd commented:
Yeah, it's always tough to start reviews on existing code, since the suggested changes inevitably extend beyond what's in the specific PR. Basically I'm trying to sort of gently introduce things that I think are a good idea as part of these reviews rather than heaping a pile of "this is the way to do things" into one big document or anything. The idea being that that idea is remembered next time and just kind of done by default.

Similarly, each change is an opportunity to clean up existing code around it rather than just trying to fit the change into the existing, less than ideal arrangement.

I also totally get that it's uncomfortable having other people ask you to change your code - I went through exactly the same thing at my last job. All I can say is that in the end the code I was producing was a lot better for it.

alhambra1 commented:
For my own part, I like changing my code when I find there's a good reason
for it.

darrenklein commented:
I'm not uncomfortable with being asked (or suggested) to change my code, that's totally fine.

bernardd commented:
Looks good, thanks. I'm going to have to take your (and the tests) word about the actual DynamoDB correctness since I'm not really across that yet :)

darrenklein commented:
Thanks for the suggestions, @bernardd , I've incorporated those changes.

Having said that, I'm not merging this PR yet, as the logic I introduced here seems like it requires further discussion, in regards to the `query` vs `get-item` behavior.

darrenklein commented:
Ah - you merged the PR? I'm not sure that we should yet - I think this issue requires some input from our fearless leader.

darrenklein commented:
Well, it works... we could always revisit it for efficiency upgrades later...

bernardd commented:
Well, the good news is that we can always un-merge it :)

bernardd commented:
I've reopened this over in #52.

bernardd commented:
Apologies - generally if I think a PR isn't ready to be merged (and requires more discussion) I try to be explicit about it, otherwise I assume that a PR is ready to be P'd :)

alhambra1 commented:
I think for me opinions here wouldn't be enough. We're looking at taking potentially many single records and instead of calling `batch_get_item` to directly retrieve those, running a single, sequential query (all in a single thread) for each item.
